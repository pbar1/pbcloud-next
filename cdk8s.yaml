language: typescript
app: yarn ts-node main.ts
imports:
- k8s
