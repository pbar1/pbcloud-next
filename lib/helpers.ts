import * as k8s from "../imports/k8s";
import * as path from "path";

/**
 * Given an image like "ghcr.io/example/img:latest", return "img".
 * @param image Container image.
 */
export function nameFromImage(image: string): string {
  return path.basename(image).split(":")[0];
}

/**
 * Convenience function for creating a Kubernetes container environment
 * variable from a name-value pair.
 * @param name Environment variable name.
 * @param value Environment variable value.
 */
export function env(name: string, value: string): k8s.EnvVar {
  return { name, value };
}

/**
 * Convenience function for creating a Kubernetes container environment
 * variable from a Kubernetes secret key.
 * @param name Environment variable name.
 * @param secret Kubernetes secret name.
 * @param key Kubernetes secret key containing environment variable value.
 */
export function envSec(name: string, secret: string, key: string): k8s.EnvVar {
  return {
    name,
    valueFrom: { secretKeyRef: { name: secret, key } },
  };
}

/**
 * Convenience function for creating a Kubernetes container port from the given
 * values.
 * @param port Port number.
 * @param protocol Port protocol.
 * @param name Port name.
 * @returns
 */
export function port(
  port: number,
  protocol?: string,
  name?: string
): k8s.ContainerPort {
  return {
    containerPort: port,
    protocol,
    name,
  };
}

/**
 * Convenience function for creating a Kubernetes volume and volume mount of
 * type `hostPath`.
 * @param name Volume name.
 * @param hostPath Path on the host.
 * @param mountPath Path within the container.
 * @param readOnly Whether to mount the path as read-only.
 * @returns A tuple containing the volume and volume mount.
 */
export function hostPath(
  name: string,
  hostPath: string,
  mountPath: string,
  readOnly?: boolean
): [k8s.Volume, k8s.VolumeMount] {
  return [
    { name, hostPath: { path: hostPath } },
    { name, mountPath, readOnly },
  ];
}
