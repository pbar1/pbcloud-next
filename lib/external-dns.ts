import { Construct } from "constructs";
import { Chart, ChartProps, Helm } from "cdk8s";
import * as path from "path";

export class ExternalDNS extends Chart {
  constructor(scope: Construct, id: string, props: ChartProps = {}) {
    super(scope, id, props);

    const chartDir = path.join(__dirname, "..", "helm", "external-dns");
    new Helm(this, "external-dns", {
      chart: chartDir,
    });
  }
}
