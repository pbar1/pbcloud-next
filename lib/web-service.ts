import { Construct } from "constructs";
import { Names } from "cdk8s";
import { Deployment, Service, IntOrString } from "../imports/k8s";

export interface WebServiceProps {
  /**
   * The container image to use for this service.
   */
  readonly image: string;

  /**
   * Number of replicas.
   *
   * @default 1
   */
  readonly replicas?: number;

  /**
   * External port.
   *
   * @default 80
   */
  readonly port?: number;

  /**
   * Internal port.
   *
   * @default 8080
   */
  readonly containerPort?: number;

  /**
   * Service type.
   *
   * @default "ClusterIP"
   */
  readonly serviceType?: string;
}

export class WebService extends Construct {
  constructor(scope: Construct, id: string, props: WebServiceProps) {
    super(scope, id);

    const port = props.port || 80;
    const containerPort = props.containerPort || 8080;
    const serviceType = props.serviceType || "ClusterIP";
    const label = { app: Names.toDnsLabel(this) };
    const replicas = props.replicas ?? 1;

    new Service(this, "service", {
      spec: {
        type: serviceType,
        ports: [{ port, targetPort: IntOrString.fromNumber(containerPort) }],
        selector: label,
      },
    });

    new Deployment(this, "deployment", {
      spec: {
        replicas,
        selector: {
          matchLabels: label,
        },
        template: {
          metadata: { labels: label },
          spec: {
            containers: [
              {
                name: "app",
                image: props.image,
                ports: [{ containerPort }],
              },
            ],
          },
        },
      },
    });
  }
}
