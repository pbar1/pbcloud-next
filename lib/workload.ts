import { Construct } from "constructs";
import * as k8s from "../imports/k8s";
import * as helpers from "./helpers";
import { Chart } from "cdk8s";
import { WritableDeep } from "type-fest";

// ----------------------------------------------------------------------------

/**
 * Creates a new container builder.
 * @param image Container image.
 * @param name Container name. If not set, defaults to the basename of the image.
 * @returns
 */
export function container(image: string, name?: string): ContainerBuilder {
  return new ContainerBuilder(image, name);
}

export class ContainerBuilder {
  private container: WritableDeep<k8s.Container>;

  constructor(image: string, name?: string) {
    name = name ?? helpers.nameFromImage(image);
    this.container = { name, image };
  }

  withEnv(env: k8s.EnvVar) {
    if (!this.container.env) {
      this.container.env = [];
    }
    this.container.env.push(env);
    return this;
  }

  withPort(port: k8s.ContainerPort) {
    if (!this.container.ports) {
      this.container.ports = [];
    }
    this.container.ports.push(port);
    return this;
  }

  asWorkload(): WorkloadBuilder {
    return new WorkloadBuilder()
      .withContainer(this.container)
      .withName(this.container.name);
  }

  build(): k8s.Container {
    return this.container;
  }
}

// ----------------------------------------------------------------------------

export enum WorkloadType {
  Deployment,
  StatefulSet,
  DaemonSet,
}

export interface WorkloadProps {
  name?: string;
  namespace?: string;
  workloadType?: WorkloadType;
  expose?: boolean;
  containers?: WritableDeep<k8s.Container>[];
  volumes?: WritableDeep<k8s.Volume>[];
}

export class WorkloadBuilder {
  private props: WorkloadProps;

  constructor() {
    this.props = {};
  }

  withName(name: string) {
    this.props.name = name;
    return this;
  }

  withNamespace(namespace: string) {
    this.props.namespace = namespace;
    return this;
  }

  withWorkloadType(workloadType: WorkloadType) {
    this.props.workloadType = workloadType;
    return this;
  }

  withExpose() {
    this.props.expose = true;
    return this;
  }

  withContainer(container: k8s.Container) {
    if (!this.props.containers) {
      this.props.containers = [];
    }
    this.props.containers.push(container);
    return this;
  }

  /**
   * Adds a volume and volume mount pair to the workload.
   * @param pair Tuple containing the volume and volume mount.
   * @param containerName Applies the volume mount to only this container.
   */
  withVolumeAndMount(
    pair: [k8s.Volume, k8s.VolumeMount],
    containerName?: string
  ) {
    let [volume, mount] = pair;

    if (!this.props.volumes) {
      this.props.volumes = [];
    }
    this.props.volumes.push(volume);

    if (!this.props.containers) {
      this.props.containers = [];
    }
    this.props.containers
      .filter((container) => !containerName ?? container.name === containerName)
      .forEach((container) => {
        if (!container.volumeMounts) {
          container.volumeMounts = [];
        }
        container.volumeMounts.push(mount);
      });

    return this;
  }

  build(scope: Construct, id: string): Workload {
    return new Workload(scope, id, this.props);
  }
}

export class Workload extends Chart {
  constructor(scope: Construct, id: string, props: WorkloadProps) {
    super(scope, id);

    if (!props.containers || props.containers.length < 1) {
      throw new Error("must have at least 1 container");
    }

    const name = props.name ?? id;
    const namespace = props.namespace;
    const workloadType = props.workloadType ?? WorkloadType.Deployment;
    const expose = props.expose ?? false;
    const containers = props.containers;
    const volumes = props.volumes;
    const selectorLabels = { app: name };

    if (workloadType === WorkloadType.Deployment) {
      new k8s.Deployment(this, "deployment", {
        metadata: { name, namespace },
        spec: {
          selector: { matchLabels: selectorLabels },
          template: {
            spec: { containers, volumes },
          },
        },
      });
    } else if (workloadType === WorkloadType.StatefulSet) {
      throw new Error("statefulset not implemented");
    } else if (workloadType === WorkloadType.DaemonSet) {
      throw new Error("daemonset not implemented");
    }

    if (expose) {
      let servicePorts: k8s.ServicePort[] = [];
      containers.forEach((ctr) => {
        ctr.ports?.forEach((port) =>
          servicePorts.push({
            name: port.name,
            protocol: port.protocol,
            port: port.containerPort,
          })
        );
      });

      new k8s.Service(this, "service", {
        metadata: { name, namespace },
        spec: {
          selector: selectorLabels,
          ports: servicePorts,
        },
      });
    }
  }
}
