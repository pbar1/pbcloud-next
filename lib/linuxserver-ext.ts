import { ContainerBuilder } from "./workload";
import { env } from "./helpers";

declare module "./workload" {
  /**
   * Adds defaults that LinuxServer.io containers expect.
   */
  interface ContainerBuilder {
    withLinuxServerDefaults(): this;
  }
}

ContainerBuilder.prototype.withLinuxServerDefaults =
  function (): ContainerBuilder {
    this.withEnv(env("TZ", "America/Los_Angeles"))
      .withEnv(env("PUID", "1000"))
      .withEnv(env("PGID", "100"));
    return this;
  };
