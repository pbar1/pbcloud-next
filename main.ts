import { App } from "cdk8s";
import { container } from "./lib/workload";
import { port } from "./lib/helpers";
import "./lib/linuxserver-ext";

const app = new App();

container("ghcr.io/linuxserver/sonarr:latest")
  .withPort(port(8989))
  .asLinuxServerWorkload()
  .withTorrentMount()
  .withTvMount()
  .withNamespace("media")
  .build(app);

container("ghcr.io/linuxserver/radarr:latest")
  .withPort(port(7878))
  .asLinuxServerWorkload()
  .withTorrentMount()
  .withMoviesMount()
  .withNamespace("media")
  .build(app);

container("ghcr.io/linuxserver/readarr:develop")
  .withPort(port(8787))
  .asLinuxServerWorkload()
  .withTorrentMount()
  .withAudiobooksMount()
  .withNamespace("media")
  .build(app);

app.synth();
