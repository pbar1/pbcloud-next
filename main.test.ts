import { ExternalDNS } from "./lib/external-dns";
import { MyChart } from "./main";
import { Testing } from "cdk8s";

describe("Placeholder", () => {
  test("Empty", () => {
    const app = Testing.app();
    const chart = new MyChart(app, "test-chart");
    const results = Testing.synth(chart);
    expect(results).toMatchSnapshot();
  });

  // TODO: Move to be inline with external-dns lib
  test("ExternalDNS", () => {
    const app = Testing.app();
    const chart = new ExternalDNS(app, "test-chart");
    const results = Testing.synth(chart);
    expect(results).toMatchSnapshot();
  });
});
