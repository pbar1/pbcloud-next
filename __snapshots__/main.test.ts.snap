// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Placeholder Empty 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "name": "test-chart-hello-k8s-service-c818d05c",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 80,
          "targetPort": 8080,
        },
      ],
      "selector": Object {
        "app": "test-chart-hello-k8s-c89248e0",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "test-chart-hello-k8s-deployment-c8a53eee",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "app": "test-chart-hello-k8s-c89248e0",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "app": "test-chart-hello-k8s-c89248e0",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "image": "paulbouwer/hello-kubernetes:1.7",
              "name": "app",
              "ports": Array [
                Object {
                  "containerPort": 8080,
                },
              ],
            },
          ],
        },
      },
    },
  },
]
`;

exports[`Placeholder ExternalDNS 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "ServiceAccount",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "test-chart-external-dns-c866fa96",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "external-dns",
        "app.kubernetes.io/version": "0.13.5",
        "helm.sh/chart": "external-dns-1.13.0",
      },
      "name": "test-chart-external-dns-c866fa96",
      "namespace": "data",
    },
  },
  Object {
    "apiVersion": "rbac.authorization.k8s.io/v1",
    "kind": "ClusterRole",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "test-chart-external-dns-c866fa96",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "external-dns",
        "app.kubernetes.io/version": "0.13.5",
        "helm.sh/chart": "external-dns-1.13.0",
      },
      "name": "test-chart-external-dns-c866fa96",
    },
    "rules": Array [
      Object {
        "apiGroups": Array [
          "",
        ],
        "resources": Array [
          "nodes",
        ],
        "verbs": Array [
          "list",
          "watch",
        ],
      },
      Object {
        "apiGroups": Array [
          "",
        ],
        "resources": Array [
          "pods",
        ],
        "verbs": Array [
          "get",
          "watch",
          "list",
        ],
      },
      Object {
        "apiGroups": Array [
          "",
        ],
        "resources": Array [
          "services",
          "endpoints",
        ],
        "verbs": Array [
          "get",
          "watch",
          "list",
        ],
      },
      Object {
        "apiGroups": Array [
          "extensions",
          "networking.k8s.io",
        ],
        "resources": Array [
          "ingresses",
        ],
        "verbs": Array [
          "get",
          "watch",
          "list",
        ],
      },
    ],
  },
  Object {
    "apiVersion": "rbac.authorization.k8s.io/v1",
    "kind": "ClusterRoleBinding",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "test-chart-external-dns-c866fa96",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "external-dns",
        "app.kubernetes.io/version": "0.13.5",
        "helm.sh/chart": "external-dns-1.13.0",
      },
      "name": "test-chart-external-dns-c866fa96-viewer",
    },
    "roleRef": Object {
      "apiGroup": "rbac.authorization.k8s.io",
      "kind": "ClusterRole",
      "name": "test-chart-external-dns-c866fa96",
    },
    "subjects": Array [
      Object {
        "kind": "ServiceAccount",
        "name": "test-chart-external-dns-c866fa96",
        "namespace": "data",
      },
    ],
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "test-chart-external-dns-c866fa96",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "external-dns",
        "app.kubernetes.io/version": "0.13.5",
        "helm.sh/chart": "external-dns-1.13.0",
      },
      "name": "test-chart-external-dns-c866fa96",
      "namespace": "data",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "name": "http",
          "port": 7979,
          "protocol": "TCP",
          "targetPort": "http",
        },
      ],
      "selector": Object {
        "app.kubernetes.io/instance": "test-chart-external-dns-c866fa96",
        "app.kubernetes.io/name": "external-dns",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "test-chart-external-dns-c866fa96",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "external-dns",
        "app.kubernetes.io/version": "0.13.5",
        "helm.sh/chart": "external-dns-1.13.0",
      },
      "name": "test-chart-external-dns-c866fa96",
      "namespace": "data",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "app.kubernetes.io/instance": "test-chart-external-dns-c866fa96",
          "app.kubernetes.io/name": "external-dns",
        },
      },
      "strategy": Object {
        "type": "Recreate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "app.kubernetes.io/instance": "test-chart-external-dns-c866fa96",
            "app.kubernetes.io/name": "external-dns",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "args": Array [
                "--log-level=info",
                "--log-format=text",
                "--interval=1m",
                "--source=service",
                "--source=ingress",
                "--policy=upsert-only",
                "--registry=txt",
                "--provider=aws",
              ],
              "image": "registry.k8s.io/external-dns/external-dns:v0.13.5",
              "imagePullPolicy": "IfNotPresent",
              "livenessProbe": Object {
                "failureThreshold": 2,
                "httpGet": Object {
                  "path": "/healthz",
                  "port": "http",
                },
                "initialDelaySeconds": 10,
                "periodSeconds": 10,
                "successThreshold": 1,
                "timeoutSeconds": 5,
              },
              "name": "external-dns",
              "ports": Array [
                Object {
                  "containerPort": 7979,
                  "name": "http",
                  "protocol": "TCP",
                },
              ],
              "readinessProbe": Object {
                "failureThreshold": 6,
                "httpGet": Object {
                  "path": "/healthz",
                  "port": "http",
                },
                "initialDelaySeconds": 5,
                "periodSeconds": 10,
                "successThreshold": 1,
                "timeoutSeconds": 5,
              },
              "securityContext": Object {
                "capabilities": Object {
                  "drop": Array [
                    "ALL",
                  ],
                },
                "readOnlyRootFilesystem": true,
                "runAsNonRoot": true,
                "runAsUser": 65534,
              },
            },
          ],
          "securityContext": Object {
            "fsGroup": 65534,
          },
          "serviceAccountName": "test-chart-external-dns-c866fa96",
        },
      },
    },
  },
]
`;
